### PR TITLE
Improve Bazel XCHammer consumption / distribution

### PR DESCRIPTION
**Pull request by jerrymarino**
This PR makes it easier to pull XCHammer into a WORKSPACE with Bazel. The
possibilities are:

*xchammer_dist_repo*
Pulled as a binary distributed artifact e.g. github release

*xchammer_dev_repo*
From a local repository, but built inside of the respective repository

*source: git_repository or local_repository*
pulling in the entire repository and building it in tree

Additionally includes multiple breaking changes to the public starlark
API
- it has to rename `xchammer_resources` to `xchammer`
- it removes the need to specify xchammer in the `xcode_project` rule.
This invariant is enforced by the fact the binary is always in the same
repo as the rules.

### CODE REVIEW

File: BUILD.bazel [hunk 1]
```diff
@@ -175,3 +189,77 @@ xcode_project(
...
+# local_repository(
+#   name = "xchammer",
+#   path =  "/path/to/xchammer/bazel-bin/xchammer_dev_repo/"
+# )
+# --override_repository=xchammer=/path/to/xchammer/bazel-bin/xchammer_dev_repo/
+genrule(
+    name = "xchammer_dev",
+    srcs = [ ":xchammer" ],
+    cmd = """
+    set -x
```
Comment (jparise) [inline BUILD.bazel:258]: Why `set -x` here and not elsewhere?

**Review by jparise — APPROVED** [review]

File: BUILD.bazel (hunk 1 above)
Comment (jerrymarino) [reply BUILD.bazel:258]: @jparise good catch, this is certainly left over from development  - will remove

**Comment by acecilia** [comment]
> This PR makes it easier to pull XCHammer into a WORKSPACE with Bazel

Very nice! Would it be possible to add a section to the documentation explaining the three alternatives in detail (`xchammer_dist_repo`, `xchammer_dev_repo` and `source`), together with some starlark code showcasing each of them?

**Comment by jerrymarino** [comment]
@acecilia cool glad you like it! I'll add update more for this, and we've got a few more API breaking changes to make as well

My current plan is
### 1 pull binary releases of XCHammer into an iOS App's build with `http_archive`
```
 http_archive(name="xchammer", urls=["https://github/../xchammer-release-{commit-sha}.zip"])
```
### 2 cut binary release builds of each commit on the github ci

### 3a develop xchammer locally in an iOS App's WORKSPACE by adding  
```
--override_repository=xchammer=/path/to/xchammer/bazel_bin/xchammer_dev_repo
````
or
### 3b  develop xchammer in an iOS App's WORKSPACE by adding
```
--override_repository=xchammer=/path/to/xchammer
````

This being said, these steps and binary releases aren't 100% necessary and I'd imagine some folks might just pull it in `git_repository`. For my setup this would bring troubles like xcode / swift versioning, bazel versioning, rule versioning, and re-builds. There's no real guidance on this from Bazel, but I personally pull tools as binary releases to prevent environment complexity and re-builds

**Comment by acecilia** [comment]
Sounds good. Would it be possible to add that information as documentation inside this PR?

**Comment by jerrymarino** [comment]
Sure, I can submit a docs PR for this as a followup once the CI workflow is landed down and we've got the exact URL scheme worked out. There's still a few details on all of this I don't know the answer to

File: BUILD.bazel [hunk 2]
```diff
@@ -175,3 +189,76 @@ xcode_project(
...
+    outs = [ "xchammer_unzip" ]
+)
+
+# This produces the same interface as the XCHammer repository, so a person
+# could use a distribution artifact or a source build without changing thier
+# build files
+# Building XCHammer ( or any decently complex tool using external deps ) from
+# source isn't suggested as it will involve making sure your iOS env has the
+# same bazel, Xcode version installed, and all transitive deps in the
+# WORKSPACE. Outside of being difficulto update/manage, it's also very
```
Comment (garrettmoon) [inline BUILD.bazel:206]: nit: typo

**Review by garrettmoon — COMMENTED** [review]
Awesome, nice work Jerry!

**Pull request merged**